src/services/completions.ts

```diff
@@ -1280,31 +1331,64 @@ namespace ts.Completions {
+                    // If `!!d.parent.parent.moduleSpecifier`, this is `export { foo } from "foo"` re-export, which creates a new symbol (thus isn't caught by the first check).
+                    if (some(symbol.declarations, d => isExportSpecifier(d) && !d.propertyName && !!d.parent.parent.moduleSpecifier)) {
+                        // Walk the export chain back one module (step 1 or 2 in diagrammed example)
+                        const nearestExportSymbol = Debug.assertDefined(getNearestExportSymbol(symbol));
+                        if (!symbolHasBeenSeen(nearestExportSymbol)) {
+                            aliasesToReturnIfOriginalsAreMissing.set(getSymbolId(nearestExportSymbol).toString(), { alias: symbol, moduleSymbol, insertAt: symbols.length });
+                            aliasesToAlreadyIncludedSymbols.set(getSymbolId(symbol).toString(), true);
+                        }
+                        else {
+                            // Perf - we know this symbol is an alias to one that’s already covered in `symbols`, so store it here
+                            // in case another symbol re-exports this one; that way we can short-circuit as soon as we see this symbol id.
+                            addToSeen(aliasesToAlreadyIncludedSymbols, getSymbolId(symbol));
+                        }
                     }
+                    else {
+                        // This is not a re-export, so see if we have any aliases pending and remove them (step 3 in diagrammed example)
+                        aliasesToReturnIfOriginalsAreMissing.delete(getSymbolId(symbol).toString());
+                        pushSymbol(symbol, moduleSymbol);
+                    }
+                }
+            });
+
+            // By this point, any potential duplicates that were actually duplicates have been
+            // removed, so the rest need to be added. (Step 4 in diagrammed example)
+            aliasesToReturnIfOriginalsAreMissing.forEach(({ alias, moduleSymbol, insertAt }) => pushSymbol(alias, moduleSymbol, insertAt));
+
+            function pushSymbol(symbol: Symbol, moduleSymbol: Symbol, insertAt = symbols.length) {
+                const isDefaultExport = symbol.escapedName === InternalSymbolName.Default;
+                if (isDefaultExport) {
+                    symbol = getLocalSymbolForExportDefault(symbol) || symbol;
+                }
+                const origin: SymbolOriginInfoExport = { kind: SymbolOriginInfoKind.Export, moduleSymbol, isDefaultExport };
+                if (detailsEntryId || stringContainsCharactersInOrder(getSymbolName(symbol, origin, target).toLowerCase(), tokenTextLowerCase)) {
+                    symbols.splice(insertAt, 0, symbol);
+                    symbolToSortTextMap[getSymbolId(symbol)] = SortText.AutoImportSuggestions;
+                    symbolToOriginInfoMap[getSymbolId(symbol)] = origin;
+                }
+            }
+
+            function symbolHasBeenSeen(symbol: Symbol) {

```

Comment:
this is only used once, so I would actually inline it. My eye had to jump around a lot to look for uses of aliasesToAlreadyIncludedSymbols, and inlining this would put them all in the same region.

It would also dedupe some calls to getSymbolId, but I imagine that is a cheap, inlinable function anyway.

Reply:
Hm, given that there are a couple different data structures here, I felt like this read a lot more clearly when given a name. I suppose I can just assign it to a const for name-clarity purposes.